### ISSUE TITLE
Visualization bugs in Plot: Contour position dependent on spacing resolution

### ISSUE THREAD

**Issue by chrherold** [DC-analysis/DCscope#110]
found the following unexpected behavior in Shape-Out 2.8.1:

The position of the contour lines seem not to match the density plot at all at some settings for spacing

<img width="673" alt="Bildschirmfoto 2021-11-12 um 14 57 58" src="https://user-images.githubusercontent.com/13899168/144019943-0849add1-6738-489f-a07c-22b0fc9e6632.png">

while they match the density plot for other spacing settings

<img width="673" alt="Bildschirmfoto 2021-11-12 um 14 58 21" src="https://user-images.githubusercontent.com/13899168/144020049-48f3cdb9-0ae6-46f4-bf44-d66ca27eb11a.png">

I can imagine that the spacing resolution influences the contour accuracy but I believe it should not completely mismatch
the contour position if the underlying density map has its maximum clearly elsewhere.

**Comment by paulmueller** [comment]
It certainly looks strange. ~Can I access that dataset somewhere? I could not reproduce it on my machine.~ I can reproduce it.

**Comment by paulmueller** [comment]
The thing is that the density map you see in the scatter plot is not identical to the density map used for contour computation.
Also note that Shape-Out computes the contour lines based on percentiles (events within a contour line). The artifacts that you are seeing are due to extremely large bins in the histogram binning method (The contour lines are generated with sub-bin-width accuracy, so to say). Here are a few examples that illustrate this:

![image](https://user-images.githubusercontent.com/2853022/144133916-ce953b71-699f-4035-b345-2d5475d1ec58.png)

![image](https://user-images.githubusercontent.com/2853022/144134117-a2c37b99-4eeb-4cbc-b58f-72f93e070449.png)

![image](https://user-images.githubusercontent.com/2853022/144134009-b4b4fe16-2b8d-4adf-9408-b69b12170604.png)

Basically, whenever you have "kinks" in your contour line, your spacing is too large.

I have the feeling that it might be reasonable to automatically refine the spacing until there are no more kinks at the expense of the user having to wait longer for their plots.

@chrherold  What do you think?

**Comment by paulmueller** [comment]
As discussed,

- a temporary solution to this issue is a button that can automatically refine the contour spacing based on kinks and a warning in the plot that states that the contour spacing is too large
- an ultimate solution based on a new implementation in dclab: https://github.com/ZELLMECHANIK-DRESDEN/dclab/issues/153 (see #111)

**Comment by paulmueller** [comment]
Bullet point 1 in my previous post is resolved with ebe507a and 88ab772.

Here is an example (note the red text):

![image](https://user-images.githubusercontent.com/2853022/144768493-ac1144e6-19e1-4c5f-bfba-a9a6cca068d0.png)

And there is a new button in the plot tab for setting "sane" values to the contour spacing:

![image](https://user-images.githubusercontent.com/2853022/144768527-7157d0bd-ffcd-4c1f-9608-48d32cec7c15.png)

@chrherold Please have a look at the development release here and let me know whether we can release this temporary solution: https://ci.appveyor.com/project/paulmueller/shapeout2/builds/41778981/artifacts


If this works well enough, I would close this issue here. Regarding the second bullet point, see issue #111.

**Comment by chrherold** [comment]
I tested the temporary solution and it works very well. Especially the find best spacing button does a very good job.

The indicator in the plot seems to move around a bit weirdly, if there is more than one row

![image](https://user-images.githubusercontent.com/13899168/145222971-9520f783-fc25-40be-b086-b95c6c177779.png)

And it shows multiple times if "One plot per dataset" is selected. Maybe it could be placed inside the plotting panels that contain a contour drawing. This way there would also not be some cutting out a part of the figure around it.

**Comment by paulmueller** [comment]
Thanks for testing.
Please try the latest dev version: https://ci.appveyor.com/project/paulmueller/shapeout2/builds/41818695/artifacts
![image](https://user-images.githubusercontent.com/2853022/145242108-a8052312-4cb3-4ced-ae15-69cd0a3afc24.png)

**Comment by chrherold** [comment]
Works very well! Thanks!
